modelscope_agent/rag/knowledge.py

```diff
@@ -0,0 +1,131 @@
+import os
+from typing import Dict, Any, Union, List, Optional
+
+from llama_index.core.llms.llm import LLM
+from llama_index.core.schema import Document
+from llama_index.core.readers.base import BaseReader
+from llama_index.core.llama_pack.base import BaseLlamaPack
+from llama_index.core.query_engine import RetrieverQueryEngine
+from llama_index.core.indices.service_context import ServiceContext
+from llama_index.core import VectorStoreIndex, SimpleDirectoryReader
+
+
+@register_rag('base_pipeline')
+class BaseKnowledgePipeline(BaseLlamaPack):
+    """ base knowledge pipeline.
+
+    从不同的源加载知识，支持：文件夹路径（str），文件路径列表（list），将不同源配置到不同的召回方式（dict）.
+    Automatically select the best file reader given file extensions.
+
+    Args:
+        knowledges: Path to the directory，或文件路径列表，或指定召回方式的文件路径。
+        save_path: 缓存indexing后的信息。
+        llm: 总结召回内容时使用的llm。
+    """
+
+    def __init__(self,
+                 knowledge_source: Union[Dict, List, str],
+                 cache_dir: str = './run',
+                 llm: Optional[LLM] = None,
+                 **kwargs) -> None:
+        extra_readers = self.get_extra_readers()
```

Comment:
what's different between readers, tools, retriever?
which part should user customize?